### PR TITLE
4. change(rpc): Refactor RPCs using a state tip height function

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We call the same code 4 times to get the state tip height, so I made a function for it.

## Review

This is a tiny refactor, anyone can review it.

### Reviewer Checklist

  - [x] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [x] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

File: zebra-rpc/src/methods.rs [hunk 1]
```diff
@@ -994,6 +992,19 @@ where
     }
 }
 
+/// Returns the best chain tip height of `latest_chain_tip`,
+/// or an RPC error if there are no blocks in the state.
+pub fn best_chain_tip_height<Tip>(latest_chain_tip: &Tip) -> Result<Height>
+where
+    Tip: ChainTip + Clone + Send + Sync + 'static,
+{
+    latest_chain_tip.best_tip_height().ok_or(Error {
```
Comment (arya2) [inline zebra-rpc/src/methods.rs:1001]: On this note it might be nice to have a `make_server_error` fn in the future.

**Review by arya2 — DISMISSED** [review]
Looks good to me, thank you for the cleanup.

File: zebra-rpc/src/methods.rs (hunk 1 above)
Comment (teor2345) [reply zebra-rpc/src/methods.rs:1001]: Something like this, but maybe with an `error_code: Into<Option<i64>>` argument.

```rust
impl<T, E> MapServerError for Result<T, E> where E: Debug {
    fn map_server_error(result: Self) -> Result<T, jsonrpc_core::Error> {
        result.map_err(|error| jsonrpc_core::Error {
            code: ErrorCode::ServerError(0),
            message: format!("{error:?}"),
            data: None,
        })
    }
}

impl<T> OkOrServerError for Option<T> {
    fn ok_or_server_error<S: ToString>(option: Self, message: S) -> Result<T, jsonrpc_core::Error> {
        option.ok_or(jsonrpc_core::Error {
            code: ErrorCode::ServerError(0),
            message: message.to_string(),
            data: None,
        })
    }
}
```

**Comment by dconnolly** [comment]
hm this may need a rebase
image

**Comment by teor2345** [comment]
> hm this may need a rebase

GitHub does this when the workflows are different for PRs targeting `main`. Seems like a CI bug we should fix.

This PR will get rebased anyway, after we fix `main` branch failures in PR #5549.

**Comment by teor2345** [comment]
> > hm this may need a rebase
> 
> GitHub does this when the workflows are different for PRs targeting `main`. Seems like a CI bug we should fix.

I opened PR #5550 which should fix this.

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
This PR conflicts with PR #5526, so I pushed a merge commit, and set it to merge after that PR merges.

**Review by arya2 — APPROVED** [review]

**Pull request merged**